File: mcdowell/python/ch11-merge-sort.py
def mergeSort(inputs):
    if len(inputs) <= 1:
        return inputs

    mid = len(inputs) // 2

    left = mergeSort(inputs[0:mid])
    right = mergeSort(inputs[mid:])
    return merge(left, right)


def merge(left, right):
    ans = []
    leftIndex = 0
    rightIndex = 0
    leftCount = len(left)
    rightCount = len(right)

    while leftIndex < leftCount and rightIndex < rightCount:
        if left[leftIndex] <= right[rightIndex]:
            ans.append(left[leftIndex])
            leftIndex += 1
        else:
            ans.append(right[rightIndex])
            rightIndex += 1

    while leftIndex < leftCount:
        ans.append(left[leftIndex])
        leftIndex += 1

    while rightIndex < rightCount:
        ans.append(right[rightIndex])
        rightIndex += 1

    return ans



print mergeSort([3,45,6543,2,3,4,5,6])

File: mcdowell/php/ch11-merge-sort.php
<?php


function mergeSort($input) {
    if (count($input) <= 1) return $input;

    $mid = (int) (count($input)/2);

    $leftData = mergeSort( array_slice($input, 0, $mid) );
    $rightData = mergeSort( array_slice($input, $mid ) );

    return merge($leftData, $rightData);
}


function merge($left, $right) {

    $leftIndex = $rightIndex = 0;
    $leftCount = count($left);
    $rightCount = count($right);
    $ans = [];
    while($leftIndex < $leftCount && $rightIndex < $rightCount) {
        if ($left[$leftIndex] <= $right[$rightIndex]) {
            $ans[] = $left[$leftIndex];
            $leftIndex ++;
        }
        else {
            $ans[] = $right[$rightIndex];
            $rightIndex ++;
        }
    }

    while($leftIndex < $leftCount) {
        $ans[] = $left[$leftIndex];
        $leftIndex ++;
    }
    while($rightIndex < $rightCount) {
        $ans[] = $right[$rightIndex];
        $rightIndex ++;
    }

    return $ans;
}


$inputs = [11,2,3,45,6,78,999];
print_r(mergeSort($inputs));